buffer.go

```diff
@@ -11,142 +10,103 @@ import (
 	"github.com/pborman/uuid"
 )
 
-// PutRecordsLimit is the maximum number of records allowed for a PutRecords request.
-var PutRecordsLimit = 500
+var ErrRecordTooBig = errors.New("data byte size is over the limit")
 
-// RecordSizeLimit is the maximum allowed size per record.
-var RecordSizeLimit int = 1 * 1024 * 1024 // 1MB
+type limits struct {
+	// PutRecordsLimit is the maximum number of records allowed for a PutRecords request.
+	putRecords int
 
-// PutRecordsSizeLimit is the maximum allowed size per PutRecords request.
-var PutRecordsSizeLimit int = 5 * 1024 * 1024 // 5MB
+	// PutRecordsSizeLimit is the maximum allowed size per PutRecords request.
+	putRecordsSize int
 
-type recordBuffer struct {
-	client   *kinesis.Kinesis
-	pKeyTmpl *template.Template
-	input    *kinesis.PutRecordsInput
-	count    int
-	byteSize int
-	mutex    sync.Mutex
+	// RecordSizeLimit is the maximum allowed size per record.
+	recordSize int
 }
 
-func newRecordBuffer(client *kinesis.Kinesis, streamName string) (*recordBuffer, error) {
-	pKeyTmpl, err := compileTmpl("KINESIS_PARTITION_KEY_TEMPLATE")
-	if err != nil {
-		return nil, err
-	}
-
-	input := &kinesis.PutRecordsInput{
-		StreamName: aws.String(streamName),
-		Records:    make([]*kinesis.PutRecordsRequestEntry, 0),
-	}
-
-	return &recordBuffer{
-		client:   client,
-		pKeyTmpl: pKeyTmpl,
-		input:    input,
-	}, nil
-}
+const (
+	PutRecordsLimit     int = 500
+	PutRecordsSizeLimit int = 5 * 1024 * 1024 // 5MB
+	RecordSizeLimit     int = 1 * 1024 * 1024 // 1MB
+)
 
-type recordSizeLimitError struct {
-	caller string
-	length int
+type buffer struct {
+	count    int
+	byteSize int
+	pKeyTmpl *template.Template
+	input    kinesis.PutRecordsInput
+	limits   *limits
 }
 
-func (e *recordSizeLimitError) Error() string {
-	return fmt.Sprintf("%s: log data byte size (%d) is over the limit.", e.caller, e.length)
+func newBuffer(tmpl *template.Template, sn string) *buffer {
+	return &buffer{
+		pKeyTmpl: tmpl,
+		input: kinesis.PutRecordsInput{
+			StreamName: aws.String(sn),
+			Records:    make([]*kinesis.PutRecordsRequestEntry, 0),
+		},
+		limits: &limits{
+			putRecords:     PutRecordsLimit,
+			putRecordsSize: PutRecordsSizeLimit,
+			recordSize:     RecordSizeLimit,
+		},
+	}
 }
 
-// Add fills the buffer with new data, or flushes it if one of the limit
-// has hit.
-func (r *recordBuffer) Add(m *router.Message) error {
-	data := m.Data
-	dataLen := len(data)
+func (b *buffer) add(m *router.Message) error {
```

Comment:
No, it's only called by the bufferMessages goroutine, and then we sent a copy of the buffer to `w.buffers`. I wonder if we could lose a message if the add() hasn't finished by the time a `reset()` is triggered by the ticker though...